File: internal/pkg/public-rpc/generate_mock_event.go
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

package publicrpc

import (
	"context"
	"errors"
	"fmt"
	"reflect"
	"zion/internal/pkg/aws/config"
	mockevent "zion/internal/pkg/mock/event"
	"zion/internal/pkg/public-rpc/types"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/schemas"
	"golang.org/x/exp/slices"
)

type GenerateMockEventsParams struct {
	RegistryName  string
	SchemaName    string
	SchemaVersion string

	EventRef     string
	Context      []string
	Overrides    string
	SkipOptional bool

	Profile string
	Region  string
}

func (p *GenerateMockEventsParams) RPCMethod() (*types.Result, error) {
	err := p.validateParams()
	if err != nil {
		return nil, err
	}

	ctx := context.TODO()
	cfg, err := config.GetAWSConfig(ctx, p.Region, p.Profile)
	if err != nil {
		return nil, fmt.Errorf("error loading AWS config: %w", err)
	}

	var schema *mockevent.Schema
	client := schemas.NewFromConfig(cfg)
	schema, err = mockevent.NewSchemaFromRegistry(
		ctx,
		aws.String(p.RegistryName),
		aws.String(p.SchemaName),
		aws.String(p.SchemaVersion),
		aws.String(p.EventRef),
		client,
	)

	if err != nil {
		return nil, fmt.Errorf("error reading schema: %w", err)
	}

	event, err := GenerateMockEvent(schema, p.SkipOptional)
	if err != nil {
		return nil, fmt.Errorf("error generating mock event: %w", err)
	}

	return &types.Result{
		Output: event,
	}, nil
}

func (p *GenerateMockEventsParams) ReflectOutput() reflect.Value {
	ft := reflect.TypeOf(GenerateMockEvent)
	out0 := ft.Out(0)
	return reflect.New(out0).Elem()
}

func (p *GenerateMockEventsParams) validateParams() error {
	if p.RegistryName == "" || p.SchemaName == "" {
		return errors.New(`requires both "RegistryName" and "SchemaName"`)
	}

	supportedContext := []string{
		"eventbridge.v0",
	}
	if p.Context != nil && len(p.Context) > 0 {
		for _, c := range p.Context {
			if !slices.Contains(supportedContext, c) {
				return fmt.Errorf("%q is not a supported context. supported context: %v", c, supportedContext)
			}
		}
	}

	return nil
}

// TODO: to be replaced by actual implementation (in internal pkg)
func GenerateMockEvent(schema *mockevent.Schema, skipOptional bool) (string, error) {

	// Commenting out as the integ tests fail, will be moved to internal pkg in a separate PR
	// generatedEvent, err := mockevent.GenerateOpenapiEvent(schema, false)

	// if err != nil {
	// 	return "", fmt.Errorf("error generating mock event: %w", err)
	// }

	// return string(generatedEvent), nil

	return "", nil
}

File: python-client/src/zion/generate_mock_event.py
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0

import json
import logging
from dataclasses import dataclass
from typing import Optional, List, Callable

from .jsonrpc import Payload


LOG = logging.getLogger(__name__)


@dataclass
class GenerateBareboneEventOutput:
    """
    zion.generate_barebone_event output
    
    Parameters
    ----------
    event : dict
        mock event
    """
    event: dict

    def __init__(self, data_dict: dict) -> None:
        event = data_dict.get("result", {}).get("output")
        self.event = json.loads(event) if event else None


@dataclass
class GenerateBareboneEventParams:
    """
    zion.generate_barebone_event parameters
    
    Parameters
    ----------
    registry_name : str
        name of the registry of the schema stored in EventBridge Schema Registry
    schema_name : str
        name of the schema stored in EventBridge Schema Registry
    schema_version : str
        version of the schema stored in EventBridge Schema Registry
    event_ref : str
        location to the event in the schema in json schema ref syntax, only applicable for openapi schema
    skip_optional : bool
        if set to true, do not generate optional fields
    """
    registry_name: Optional[str] = None
    schema_name: Optional[str] = None
    schema_version: Optional[str] = None
    event_ref: Optional[str] = None
    skip_optional: Optional[bool] = None

    _rpc_method: str = "mock.generate_barebone_event"

    def to_dict(self) -> dict:
        params = {}
        if self.registry_name:
            params["RegistryName"] = self.registry_name
        if self.schema_name:
            params["SchemaName"] = self.schema_name
        if self.schema_version:
            params["SchemaVersion"] = self.schema_version
        if self.event_ref:
            params["EventRef"] = self.event_ref
        if self.skip_optional:
            params["SkipOptional"] = self.skip_optional
        return params

    def to_payload(self, region, profile) -> Payload:
        return Payload(self._rpc_method, self.to_dict(), region, profile)


# NOTE (hawflau): client method output
@dataclass
class GenerateMockEventOutput:
    """
    zion.generate_mock_event output
    
    Parameters
    ----------
    event : dict
        mock event
    """
    event: dict


# NOTE (hawflau): client method param
@dataclass
class GenerateMockEventParams:
    """
    zion.generate_mock_event parameters

    Parameters
    ----------
    registry_name : str
        name of the registry of the schema stored in EventBridge Schema Registry
    schema_name : str
        name of the schema stored in EventBridge Schema Registry
    schema_version : str
        version of the schema stored in EventBridge Schema Registry
    event_ref : str
        location to the event in the schema in json schema ref syntax, only applicable for openapi schema
    skip_optional : bool
        if set to true, do not generate optional fields
    contexts : List[Callable[[dict], dict]]
        a list of callables to apply context on the generated mock event
    """
    registry_name: Optional[str] = None
    schema_name: Optional[str] = None
    schema_version: Optional[str] = None
    event_ref: Optional[str] = None
    skip_optional: Optional[bool] = None
    contexts: Optional[List[Callable[[dict], dict]]] = None
